### PR TITLE
Defiler primordial: tentacle

### PR DESCRIPTION
**Pull request by TiviPlus**
Grab any marine or object and pull them toward you

5 tile range, has a .5 delay between grab and throw

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm [hunk 1]
```diff
@@ -381,6 +381,67 @@ GLOBAL_LIST_INIT(defile_purge_list, typecacheof(list(
...
+	while((current != target_turf))
+		if(current.density)
+			if(!silent)
+				to_chat(owner, span_warning("We can't reach [target]!"))
+			return FALSE
+		current = get_step_towards(current, target_turf)
+
+
+/datum/action/xeno_action/activable/tentacle/use_ability(atom/movable/target)
+	var/datum/beam/tentacle = owner.beam(target,"curse0",'icons/effects/beam.dmi')
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:429]: video of the effect?

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm [hunk 2]
```diff
@@ -381,6 +381,67 @@ GLOBAL_LIST_INIT(defile_purge_list, typecacheof(list(
...
+	succeed_activate()
+	add_cooldown()
+
+///after dramatic pause throws the target at the defiler
+/datum/action/xeno_action/activable/tentacle/proc/finish_grab(atom/movable/grabbed, datum/beam/tentacle)
+	grabbed.throw_at(owner, TENTACLE_ABILITY_RANGE, 1, owner, FALSE)
+	if(isliving(grabbed))
+		var/mob/living/loser = grabbed
+		loser.apply_effects(1, 0.1)
+	qdel(tentacle)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:443]: i think there's a harddel with beams (visual not cleaned), can you fix it?

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm (hunk 1 above)
Comment (KaleRobe) [reply code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:429]: So like the viper's tongue pull from Xcom 2?

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm (hunk 2 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:443]: A) "I think"
B) out of scope

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm (hunk 1 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:429]: sent in discord

**Review by BraveMole — CHANGES_REQUESTED** [review]
Honestly the graphical effect is sub-par.

At the strict minimum, the tentacle should stick to the victim while it's being thrown toward the defiler.

The sprite itself is meh, but i don't know if it can really be improved on

It would be better if the tentacle was really reaching out from the defiler to the victim, rather than being instantly created between them as a beam

**Review by Kuro020 — APPROVED** [review]

**Comment by David-Stormwell** [comment]
Good idea, only thing I see a problem is: **If I hear a fucking slime tentacle sound affect I'm going to fucking murder you.**
Also, perhaps input more info in the PR?

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm [hunk 3]
```diff
@@ -381,6 +381,75 @@ GLOBAL_LIST_INIT(defile_purge_list, typecacheof(list(
...
+	succeed_activate()
+	add_cooldown()
+
+///after dramatic pause throws the target at the defiler
+/datum/action/xeno_action/activable/tentacle/proc/finish_grab(atom/movable/grabbed, datum/beam/tentacle)
+	RegisterSignal(grabbed, COMSIG_MOVABLE_IMPACT, .proc/delete_beam)
+	grabbed.throw_at(owner, TENTACLE_ABILITY_RANGE, 1, owner, FALSE)
+	if(isliving(grabbed))
+		var/mob/living/loser = grabbed
+		loser.apply_effects(1, 0.1)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:445]: don't use apply_effects, use the stun and other thing . This proc is way too cryptic

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm [hunk 4]
```diff
@@ -381,6 +381,75 @@ GLOBAL_LIST_INIT(defile_purge_list, typecacheof(list(
...
+	to_chat(owner, span_warning("We grab [target] with a tentacle!"))
+	target.balloon_alert_to_viewers("Grabbed!")
+	addtimer(CALLBACK(src, .proc/finish_grab, target, tentacle), 5)
+	playsound(target, 'sound/effects/blobattack.ogg', 40, 1)
+	succeed_activate()
+	add_cooldown()
+
+///after dramatic pause throws the target at the defiler
+/datum/action/xeno_action/activable/tentacle/proc/finish_grab(atom/movable/grabbed, datum/beam/tentacle)
+	RegisterSignal(grabbed, COMSIG_MOVABLE_IMPACT, .proc/delete_beam)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:441]: probably need to add a failsafe timer, unless you are 100% sure COMSIG_MOVABLE_IMPACT happens all the time

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm (hunk 3 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:445]: or do stun = 1, weaken = 0.1 eventually

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm (hunk 4 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:441]: it does its always called on throw end

File: code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm (hunk 3 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/defiler/abilities_defiler.dm:445]: done

**Review by BraveMole — APPROVED** [review]

**Pull request merged**